libz-rs-sys/src/gz.rs

```diff
@@ -753,6 +812,188 @@ pub unsafe extern "C-unwind" fn gzdirect(file: gzFile) -> c_int {
+/// to detect this case.
+///
+/// If the unsigned value `len` is too large to fit in the signed return type
+/// `c_int`, then nothing is read, `-1` is returned, and the error state is set to
+/// `Z_STREAM_ERROR`.
+///
+/// # Returns
+///
+/// * The number of uncompressed bytes read from the file into `buf`, which may
+///   be smaller than `len` if there is insufficient data in the file.
+/// * `-1` on error.
+///
+/// # Arguments
+///
+/// * `file` - A gzip file handle, or null.
+/// * `buf` - Buffer where the read data should be stored. The caller retains ownership of this buffer.
+/// * `len` - Number of bytes to attempt to read into `buf`.
+///
+/// # Safety
+///
+/// The caller must ensure that `buf` points to at least `len` writable bytes.
+#[cfg_attr(feature = "export-symbols", export_name = crate::prefix!(gzread))]
+pub unsafe extern "C-unwind" fn gzread(file: gzFile, buf: *mut c_void, len: c_uint) -> c_int {
+    let Some(state) = (unsafe { file.cast::<GzState>().as_mut() }) else {
+        return -1;
+    };
+
+    // Check that we're reading and that there's no (serious) error.
+    if state.mode != GzMode::GZ_READ || (state.err != Z_OK && state.err != Z_BUF_ERROR) {
+        return -1;
+    }
+
+    // Check that the requested number of bytes can be represented by the return type.
+    if c_int::try_from(len).is_err() {
+        // Safety: we confirmed above that state is valid.
+        unsafe {
+            gz_error(
+                state,
+                Some((Z_STREAM_ERROR, "request does not fit in an int")),
+            )
+        };
+        return -1;
+    }
+
+    // With the initial checks passed, try the actual read.
+    // Safety: The caller is responsible for ensuring that `buf` points to >= `len` writable bytes.
+    let got = unsafe { gz_read(state, buf.cast::<u8>(), len as usize) };
+
+    // Check for an error
+    if got == 0 && state.err != Z_OK && state.err != Z_BUF_ERROR {
+        -1
+    } else {
+        got as _
+    }
+}
+
+// Attempt to read enough bytes from the input to fill the supplied `buf`.
+//
+// # Returns
+//
+// The number of bytes read into `buf`. Note: A return value of zero means either end-of-file
+// or an error. `state.err` must be consulted to determine which.
+//
+// # Safety
+//
+// * `state` must have been properly initialized, e.g. by [`gzopen_help`].
+// * `buf` must point to at least `len` bytes of writable memory.
+unsafe fn gz_read(state: &mut GzState, mut buf: *mut u8, mut len: usize) -> usize {
+    if len == 0 {
+        return 0;
+    }
+
+    /* FIXME: Uncomment this when seek support is added:
+
+    // process a skip request
+    if state.seek {
+        state.seek = false;
+        if gz_skip(state, state.skip) == -1 {
+            return 0;
+        }
+    }
+     */
+
+    // Loop until we get enough bytes or reach the end of the file.
+    let mut got = 0;
+    let mut count = 0;
+    loop {
+        count += 1;
+        assert_ne!(count, 100);
```

Comment:
Is this left from debugging?

Reply:
Oops! Leftover from debugging. Thanks for catching that!